File: app/src/main/java/it/unical/mat/embasp/platforms/desktop/DesktopHandler.java
package it.unical.mat.embasp.platforms.desktop;

import java.util.List;

import it.unical.mat.embasp.base.Callback;
import it.unical.mat.embasp.base.Handler;
import it.unical.mat.embasp.base.InputProgram;
import it.unical.mat.embasp.base.OptionDescriptor;
import it.unical.mat.embasp.base.Output;

/**
 * is a specialization for a Desktop platform
 *
 * @see Handler
 */

public class DesktopHandler extends Handler {

	private final DesktopService service;

	public DesktopHandler(final DesktopService service) {
		this.service = service;
	}

	/** @see it.unical.mat.embasp.base.Service#startAsync(Callback, List, List) */
	@Override
	public void startAsync(final Callback c, final List<Integer> program_index, final List<Integer> option_index) {

		final List<InputProgram> input_programs = collect_programs(program_index);
		final List<OptionDescriptor> input_options = collect_options(option_index);

		service.startAsync(c, input_programs, input_options);

	}

	/** @see it.unical.mat.embasp.base.Service#startAsync(Callback, List, List) */
	@Override
	public Output startSync(final List<Integer> program_index, final List<Integer> option_index) {

		final List<InputProgram> input_programs = collect_programs(program_index);
		final List<OptionDescriptor> input_options = collect_options(option_index);

		return service.startSync(input_programs, input_options);
	}
}

File: CSharp version/EmbASP/it/unical/mat/embasp/platforms/desktop/DesktopHandler.cs
using System.Collections.Generic;

namespace it.unical.mat.embasp.platforms.desktop
{

	using Callback = it.unical.mat.embasp.@base.Callback;
	using Handler = it.unical.mat.embasp.@base.Handler;
	using InputProgram = it.unical.mat.embasp.@base.InputProgram;
	using OptionDescriptor = it.unical.mat.embasp.@base.OptionDescriptor;
	using Output = it.unical.mat.embasp.@base.Output;

	public class DesktopHandler : Handler
	{

		private readonly DesktopService service;

		public DesktopHandler(DesktopService service)
		{
			this.service = service;
		}

		public override void startAsync(Callback c, IList<int> program_index, IList<int> option_index)
		{

			IList<InputProgram> input_programs = collect_programs(program_index);
			IList<OptionDescriptor> input_options = collect_options(option_index);

			service.startAsync(c, input_programs, input_options);

		}

		public override Output startSync(IList<int> program_index, IList<int> option_index)
		{

			IList<InputProgram> input_programs = collect_programs(program_index);
			IList<OptionDescriptor> input_options = collect_options(option_index);

			return service.startSync(input_programs, input_options);
		}
	}

}